source/server/http/admin.cc

```diff
@@ -686,6 +686,20 @@ Http::Code AdminImpl::handlerMemory(absl::string_view, Http::HeaderMap& response
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerDrainListeners(absl::string_view url, Http::HeaderMap&,
+                                            Buffer::Instance& response, AdminStream&) {
+  const Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
+  const bool inbound_only = params.find("inboundonly") != params.end();
+  for (const Network::ListenerConfig& listener : server_.listenerManager().listeners()) {
+    if (!inbound_only || listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
+      server_.listenerManager().removeListener(listener.name());
```

Comment:
In thinking about this more, I'm not convinced this is correct. I think if you remove the listener here, it's possible that LDS will add it back? Don't we want to stop the listener instead? WDYT?

Reply:
I think assuming that it is operationally handled that this is invoked on SIGTERM and wait for drain time is done and then Envoy goes away any ways, it may not be a big issue because till the listener is draining it will not allow to add back?

Are you suggesting, we call the drain completion with empty call back here https://github.com/envoyproxy/envoy/blob/master/source/server/listener_manager_impl.cc#L668 for this case so that it won't remove?

Let me know if you want to handle that case - we will have to introduce a new API to listener manager called stopListener and call drain with null calback? or any other better way?

Reply:
Went ahead and added api for `stopListener`. Please see if this better.